### DISCUSSION TITLE
Making it generic over Arc/Rc?

### DISCUSSION THREAD

**Discussion by nyanpasu64** [Logicalshift/flo_binding#4] [Ideas]
I've observed that "traditional" GUI frameworks (eg. Qt) expect widgets to only be manipulated from the main thread. And I suspect js/wasm doesn't support threads at all.

As a result, it may be worthwhile to add non-thread-safe versions of flo_binding, built around Rc<RefCell> instead of Arc<Mutex>. Rc may be slightly faster than Arc when cloning, and RefCell is almost free, whereas Mutex's cost depends on the operating system. This also may remove the need for Send or Sync, though I'm not sure if it affects users or not. (It's been a while since I last looked at this code, so I'm a bit fuzzy on the details.)

Though Rust does not provide a good way to abstract over Rc vs. Arc, let alone RefCell vs. Mutex (which probably expose considerably different APIs). So it may not be worth the duplication and maintenance costs. (Also keep in mind I've not written an app in flo_binding, and am mostly approaching it as a source of ideas.)

**Comment by Logicalshift** [comment]
I encountered this problem myself, and thought about this approach: as you said, the problem it has is that it means duplicating most of the functionality of the library, and I found it very useful to not have to worry about whether or not a particular piece of state was single-threaded or multi-threaded. What I wound up doing instead is making the `follow()` function: it's fairly easy to write a futures executor that wakes the UI thread, and there are other benefits. Almost all UI frameworks are going to have a way to wake the UI thread, and I think in the future a built-in executor will likely be common.

What I'm assuming you want to do is something like `some_binding.when_changed(notify(|| update_ui_element()))`. Early versions of FlowBetween worked this way, but this only works if the notify function is `Send` and the callback can happen from anywhere which makes managing state harder, so usually all the notify function did was mark some part of the UI as invalid and have it update that later on.

This can also be written as `let some_future = follow(some_binding).for_each(|new_value| update_ui_element(new_value))`, then poll `some_future` on your UI thread. An immediately apparent advantage of this approach is the new value of the binding is always available from the stream. Futures can bind to a single thread, so the function here does not need to be `Send`.

A less-apparent advantage is that the update here is deferred, while `when_changed` is executed immediately on set. This means that there's no re-entrancy type problems to deal with and also that the UI generally only receives a single update per event, after the event has processed, no matter how many times any given piece of state was invalidated. `follow` streams don't return intermediate state. That's useful if there are bindings that can update multiple times during an event (only the last binding is ever actually sent to the UI), and it's especially useful if you have a computed binding: its value will only be computed a single time, even if multiple dependencies changes their value.

This also applies to UI stalls, where the rendering or an event takes a long time and the UI is stuck for a while, it will only be asked to update to the latest state after handling the pending events.

`flo_draw` has an example of this: glutin can't update its state from other threads, so I've implemented things like window properties as bindings. [Here](https://github.com/Logicalshift/flo_draw/blob/e67fecc43b804c3f0038727fa084483754a8c0db/draw/src/glutin_runtime.rs#L379) is the waker and future's runtime - any UI framework that has a function to wake from another thread can use the same structure. The future that handles the bindings is [over here](https://github.com/Logicalshift/flo_draw/blob/e67fecc43b804c3f0038727fa084483754a8c0db/draw/src/glutin_window.rs#L43) - I wrote a custom stream here instead of using `for_each`, but the principle is the same. The [update_title](https://github.com/Logicalshift/flo_draw/blob/v0.3/draw/examples/update_title.rs) example shows the whole thing in action.

**Reply by nyanpasu64** [reply]
## Thread safety

I'm also using multithreading, but in my case, the worker thread generates audio and needs to be wait-free (can't block on mutexes), so I'm going to use custom atomic-based code (effectively reimplementing https://lib.rs/crates/triple_buffer in C++) for communication. Unfortunately this won't allow the audio thread to "signal" the GUI thread about changes, and the GUI needs to poll for updates. It's not a big deal for real-time changing visualization though.

## Lazy reevaluation

> `follow` streams don't return intermediate state.

Is that true? I've heard people complaining that ReactiveX stream combinators [expose intermediate state](https://news.ycombinator.com/item?id=7966689), in a phenomenon known to the FRP community as "glitches". (But keep in mind that I know neither ReactiveX nor streams/futures.)

Since you brought up lazy reevaluation, I thought it would be appropriate to ramble. I found out to my disappointment that QProperty also relies on eager callbacks:

<blockquote><details><summary>Description of eager callbacks and glitches</summary>
Imagine a spreadsheet-ish thing:

- variable := 1
- left = variable
- right = variable
- out = left + right

and you write a callback whenever `out` changes value.

- if you don't query out's value after setting the binding, the value never gets known, and neither do the dependencies, and out never gets computed when the dependencies change, and the listener callback never gets called. so you need to add a dummy `out.value();` call.
- if you set `variable := 2`, the callback gets called once when `out = 3`, then another time when `out = 4` (QProperty has **glitches**).
</details></blockquote>

One alternative is for all calls to get queued up in the Qt event loop (rather than being invoked immediately) so the callback only runs once at the end, after all dependencies have been recomputed. (This basically treats each method call in the Qt event loop as a "transaction" of sorts.) Though I found a flaw that results from lazy reevaluation occurring in the wrong order relative to user/GUI events:

> This also applies to UI stalls, where the rendering or an event takes a long time and the UI is stuck for a while, it will only be asked to update to the latest state after handling the pending events.

You treat that as a feature, but I'm worried about a system that will process extra user inputs between the first user input and the GUI's response to it. You might get an issue much like Windows 7's installer, where when creating an account, you can type a password hint (to enable the Next button), delete it (which disables the Next button with a time delay), and click Next before it gets disabled.

Another idea I've implemented (for reconfiguring audio generators rather than a GUI) is a "mutator" object which is the only way to perform changes to state, queues up all attempted changes as `Option<T>`, and then applies them by calling the correct methods in the correct order (since it's order-dependent). Though I don't know if you can build a "mutator" object for a graph of GUI widgets with state scattered about.

----

If I read correctly, you're saying that eager evaluation is by design, and you should use streams if you want lazy reevaluation.

- I've been reading about FRP (not ReactiveX) before (though I don't understand their code almost invariably written in Haskell). [Their advocates argue](http://lambda-the-ultimate.org/node/4982) that "continuously varying values" are better than "events", and IMO it's more natural to represent a property as a time-varying value than converting it to an event stream (which ReactiveX does, and your stream feature may do as well). (There was debate over how to naturally to model user inputs as events.)
  - I haven't used ReactiveX or flo_binding's streams so I don't know if it has issues in practice. But reading [RxCpp example code](https://github.com/ReactiveX/RxCpp#example) feels like a monolithic blob of combinators, requiring you to structure your entire code around it. Qt doesn't have an API that treats event streams as objects (instead, signals are autogenerated methods and slots are user-generated methods, and QProperty is distinct from signals and slots).

The problem is mostly that I don't know how to use futures. Honestly I think I'm going to continue exploring ad-hoc reactive application design in Qt (since I already have an application written in it), without formal property binding (since QProperty is only found in the new incomplete buggy Qt 6). What libraries/frameworks do you recommend for testing out flo_binding (callbacks or streams) when building a real-world GUI app?

I have a real-world Qt application where user input can both mutate the document and move the cursor. Naively translating these to properties would result in each one of these triggering a redraw. I'm planning to instead schedule a single callback in the event loop, then write to a bitfield storing which portions of the document need to be recomputed, which the callback eventually reads. (I don't know if each user operation will cause only the user event handler to write to the bitfield, or every single leaf method that gets called.)

**Comment by Logicalshift** [comment]
So I've been quiet for a while but I have actually been thinking about this. When I've used flo_binding in my own projects that have these issues, the core UI loop has a fairly specific design which helps prevent a lot of the issues you're talking about: it's basically -

* when woken by an event: suspend UI updates and collect as many user events as possible
* process user events and update state, collect a list of which bindings have changed
* once all the events are processed, resume UI updates and process the list of changed bindings to display the new UI

That is, the idea is lazy rather than eager evaluation: bindings are not recomputed until they're actually needed: essentially sampling the state as late as possible in order to update the UI.

In a single-threaded application, you can get quite close to this by queuing manual events to cause updates when bindings notify as those will get added after any other waiting events. In multi-threaded applications (which is what FlowBetween is), I tend to use an XOn/XOff type mechanism where the UI thread has start/stop events to indicate when it's busy so that redraws don't appear too early.

You can almost get away with no background events this way: provided every update occurs in response to a user action of some kind, the delayed 'update' stage always happens at the right time to redraw everything. Right now, for background events, I've either used an 'update as soon as possible' approach or the XOn/XOff mechanism.

This prevents a couple of problematic scenarios: delaying changing the UI until just before the point the user regains control ensures that only one redraw is ever performed, and alternating between 'updating state' and 'redrawing UI' means that intermediate states are avoided.

(In case of long-running updates, it's possible to sample the state regularly to keep the UI updated, or do something like grey out controls that have been marked invalidated but not yet updated, similar to how Windows will grey out an unresponsive window)

So I think this is closest to the FRP idea of sampling. Streams vs notifications are maybe a bit of a red herring: I think the key thing to know there is that because Rust's streams use what might be thought of a hybrid polling mechanism, you can use them to implement both lazy and eager evaluation and that cuts down on the book-keeping needed (and is surprisingly intuitive once the idea is clear).

One thing I've been thinking about is that as complexity grows, it becomes harder to get to a point where you know that the state isn't being updated. To avoid displaying intermediate states, there needs to be a mechanism to find a 'safe point' where the binding values can be read knowing they are in a consistent state. `flo_binding` doesn't help here yet, but I was thinking of adding some kind of a Guard structure that could be attached to one or more bindings to indicate they're part of a common state: the UI (and things like computed bindings) could use this to ensure that only self-consistent states are generated.

That is, there'd be a way to lock and release a guard to indicate that the state is being changed, and another way to do so to indicate that the state is being read, so it's impossible to accidentally display an intermediate state. I think there could also be a notification mechanism that would notify when the guard is released and some of the bindings have changed (so it's easier to avoid blocking when trying to sample the values)

On to the original topic: I've started to think a bit more about how binding non-Send/Sync types could work, though I don't have a concrete plan yet - I'm mainly trying to avoid `flo_binding` having a parallel universe of separate implementations, though I suspect something like `bind_local()` might be inevitable.

**Comment by Logicalshift** [comment]
Oops, just spotted that your idea of Mutators is quite similar to my thought about Guards, I think maybe looking at the same issue from different angles. I managed to miss that you'd mentioned that on a second readthrough: definitely didn't intend to blank that out.
